<a href="https://colab.research.google.com/github/onlyabhilash/NLP-Code/blob/main/part-10/01_AutoML_Demo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this notebook we will demonstrate Auto ML on the mnist digits dataset. More information about autosklearn and autokeras can be found in the official documents [here](https://automl.github.io/auto-sklearn/master) and [here](https://github.com/keras-team/autokeras) respectively.<br>
Note: Auto-sklearn is not available on windows or on MAC OSX.

In [ ]:
!sudo apt-get install build-essential swig
!curl https://raw.githubusercontent.com/automl/auto-sklearn/master/requirements.txt | xargs -n 1 -L 1 pip install
!pip install auto-sklearn

Reading package lists... Done
Building dependency tree       
Reading state information... Done
build-essential is already the newest version (12.4ubuntu1).
swig is already the newest version (3.0.12-1).
0 upgraded, 0 newly installed, 0 to remove and 39 not upgraded.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   209  100   209    0     0   6531      0 --:--:-- --:--:-- --:--:--  6531


### **Auto-Sklearn** on mnist dataset

In [ ]:
import autosklearn.classification

import sklearn.datasets
import sklearn.metrics

import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split

In [ ]:
#load the data
digits = sklearn.datasets.load_digits()
X = digits.data
y = digits.target
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, random_state=1)

#initialize the classifier
automl = autosklearn.classification.AutoSklearnClassifier()
automl.fit(X_train, y_train)
y_pred = automl.predict(X_test)

#accuracy
print("Accuracy score", sklearn.metrics.accuracy_score(y_test, y_pred))


Accuracy score 0.9888888888888889


We achived an accuracy of 98.9%.<br>
Now lets understand what exactly happened there.

In [ ]:
print("Model\n",automl.show_models())
print("-")
print("Model Statistics\n",automl.sprint_statistics())

Model
 [(0.060000, SimpleClassificationPipeline({'balancing:strategy': 'weighting', 'classifier:__choice__': 'libsvm_svc', 'data_preprocessing:categorical_transformer:categorical_encoding:__choice__': 'one_hot_encoding', 'data_preprocessing:categorical_transformer:category_coalescence:__choice__': 'no_coalescense', 'data_preprocessing:numerical_transformer:imputation:strategy': 'mean', 'data_preprocessing:numerical_transformer:rescaling:__choice__': 'normalize', 'feature_preprocessor:__choice__': 'polynomial', 'classifier:libsvm_svc:C': 24427.51442319513, 'classifier:libsvm_svc:gamma': 0.0003734294061388043, 'classifier:libsvm_svc:kernel': 'sigmoid', 'classifier:libsvm_svc:max_iter': -1, 'classifier:libsvm_svc:shrinking': 'True', 'classifier:libsvm_svc:tol': 0.020946604527726176, 'feature_preprocessor:polynomial:degree': 3, 'feature_preprocessor:polynomial:include_bias': 'False', 'feature_preprocessor:polynomial:interaction_only': 'True', 'classifier:libsvm_svc:coef0': -0.1070742945645

### **Auto-Keras** on mnist dataset

In [ ]:
# Download the necessary packages. You might have to restart your runtime after this 
!pip install autokeras==1.0.2
!pip install tensorflow==2.1.0
!pip install keras-tuner==1.0.1

Processing /root/.cache/pip/wheels/bb/a1/8a/7c3de0efb3707a1701b36ebbfdbc4e67aedf6d4943a1f463d6/keras_tuner-1.0.2-cp37-none-any.whl
  Found existing installation: keras-tuner 1.0.0
    Uninstalling keras-tuner-1.0.0:
      Successfully uninstalled keras-tuner-1.0.0
     |████████████████████████████████| 61kB 5.7MB/s 
  Created wheel for keras-tuner: filename=keras_tuner-1.0.1-cp37-none-any.whl size=73199 sha256=62bc7c1cce031f059619402a5fc4133cf9a8698cee6ebcd6082146c0fe4bacb8
  Stored in directory: /root/.cache/pip/wheels/b9/cc/62/52716b70dd90f3db12519233c3a93a5360bc672da1a10ded43
Successfully built keras-tuner
  Found existing installation: keras-tuner 1.0.2
    Uninstalling keras-tuner-1.0.2:
      Successfully uninstalled keras-tuner-1.0.2


In [ ]:
from tensorflow.keras.datasets import mnist
import autokeras as ak
import warnings
warnings.filterwarnings('ignore')

In [ ]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
print(x_train.shape)
print(y_train.shape)
print(y_train[:3]) 

(60000, 28, 28)
(60000,)
[5 0 4]


In [ ]:
# split=2000

# x_test = x_test[:split]
# y_test = y_test[:split]

# x_train = x_train[:split]
# y_train = y_train[:split]


In [ ]:
# Initialize the image classifier.
clf = ak.ImageClassifier(max_trials=10) # It tries 10 different models.
# Feed the image classifier with training data.
clf.fit(x_train,y_train, epochs=3)

Train for 1500 steps, validate for 375 steps
Epoch 1/3
1500/1500 [==============================] - 9s 6ms/step - loss: 0.1764 - accuracy: 0.9460 - val_loss: 0.0666 - val_accuracy: 0.9810
Epoch 2/3
1500/1500 [==============================] - 8s 5ms/step - loss: 0.0799 - accuracy: 0.9757 - val_loss: 0.0565 - val_accuracy: 0.9827
Epoch 3/3
1500/1500 [==============================] - 8s 5ms/step - loss: 0.0636 - accuracy: 0.9801 - val_loss: 0.0516 - val_accuracy: 0.9852


Train for 1500 steps, validate for 375 steps
Epoch 1/3
1500/1500 [==============================] - 68s 46ms/step - loss: 0.2376 - accuracy: 0.9351 - val_loss: 0.1227 - val_accuracy: 0.9681
Epoch 2/3
1500/1500 [==============================] - 64s 42ms/step - loss: 0.1281 - accuracy: 0.9680 - val_loss: 0.1519 - val_accuracy: 0.9637
Epoch 3/3
1500/1500 [==============================] - 63s 42ms/step - loss: 0.0949 - accuracy: 0.9742 - val_loss: 0.2995 - val_accuracy: 0.9153


Train for 1500 steps, validate for 375 steps
Epoch 1/3
1500/1500 [==============================] - 8s 6ms/step - loss: 0.1686 - accuracy: 0.9479 - val_loss: 0.0688 - val_accuracy: 0.9814
Epoch 2/3
1500/1500 [==============================] - 8s 5ms/step - loss: 0.0772 - accuracy: 0.9766 - val_loss: 0.0524 - val_accuracy: 0.9847
Epoch 3/3
1500/1500 [==============================] - 8s 5ms/step - loss: 0.0615 - accuracy: 0.9808 - val_loss: 0.0508 - val_accuracy: 0.9848


INFO:tensorflow:Oracle triggered exit
Train for 1875 steps, validate for 375 steps
Epoch 1/3
1875/1875 [==============================] - 11s 6ms/step - loss: 0.1628 - accuracy: 0.9494 - val_loss: 0.0394 - val_accuracy: 0.9883
Epoch 2/3
1875/1875 [==============================] - 10s 5ms/step - loss: 0.0744 - accuracy: 0.9780 - val_loss: 0.0266 - val_accuracy: 0.9918
Epoch 3/3
1875/1875 [==============================] - 10s 5ms/step - loss: 0.0600 - accuracy: 0.9815 - val_loss: 0.0187 - val_accuracy: 0.9952


In [ ]:
#predict label
predicted_y = clf.predict(x_test)
print(predicted_y)  

[[7]
 [2]
 [1]
 ...
 [4]
 [5]
 [6]]


In [ ]:
#evaluate on test data
print(clf.evaluate(x_test, y_test))

313/313 [==============================] - 1s 3ms/step - loss: 0.0386 - accuracy: 0.9865
[0.03864423465912374, 0.9865]
